storage/scan_queue.go

```diff
@@ -118,15 +119,124 @@ func (sq *scanQueue) shouldQueue(now time.Time, rng *Range) (shouldQ bool, prior
 // intents are older than intentAgeThreshold. The very act of scanning
 // keys verifies on-disk checksums, as each block checksum is checked
 // on load.
-func (sq *scanQueue) process(now time.Time, rng *Range) error {
+func (sq *scanQueue) process(now proto.Timestamp, rng *Range) error {
 	snap := rng.rm.Engine().NewSnapshot()
 	iter := newRangeDataIterator(rng, snap)
 	defer iter.Close()
 	defer snap.Stop()
 
+	// Lookup the GC policy for the zone containing this key range.
+	policy, err := sq.lookupGCPolicy(rng)
+	if err != nil {
+		return err
+	}
+
+	scanMeta := proto.NewScanMetadata(now.WallTime)
+	gc := engine.NewGarbageCollector(now, policy)
+
+	// Compute intent expiration (intent age at which we attempt to resolve).
+	intentExp := now
+	intentExp.WallTime -= intentAgeThreshold.Nanoseconds()
+
+	gcArgs := &proto.InternalGCRequest{
+		RequestHeader: proto.RequestHeader{
+			Key:       rng.Desc.StartKey,
+			Timestamp: now,
+			RaftID:    rng.Desc.RaftID,
+		},
+	}
+	var expBaseKey proto.Key
+	var keys []proto.EncodedKey
+	var vals [][]byte
 	for ; iter.Valid(); iter.Next() {
-		// TODO(spencer): implement processing.
+		baseKey, ts, isValue := engine.MVCCDecodeKey(iter.Key())
+		if !isValue {
+			// Moving to the next key (& values). If there's more than a
+			// single value for the key, possibly send for GC.
+			if len(keys) > 1 {
+				meta := &proto.MVCCMetadata{}
+				if err := gogoproto.Unmarshal(vals[0], meta); err != nil {
+					log.Errorf("unable to unmarshal MVCC metadata for key %q: %s", keys[0], err)
+				} else {
+					// In the event that there's an active intent, send for
+					// intent resolution if older than the threshold.
+					startIdx := 1
+					if meta.Txn != nil {
+						if meta.Txn.Timestamp.Less(intentExp) {
+							// TODO(spencer): push txn / resolve intent. Don't forget to
+							// update the oldest intent nanos if the resolve fails.
+						} else if scanMeta.OldestIntentNanos == nil || meta.Timestamp.WallTime < *scanMeta.OldestIntentNanos {
+							scanMeta.OldestIntentNanos = gogoproto.Int64(meta.Timestamp.WallTime)
+						}
+						// With an active intent, GC ignores MVCC metadata & intent value.
+						log.Infof("there's an intent on key %q", expBaseKey)
+						startIdx = 2
+					}
+					// See if any values may be GC'd.
+					if gcTS := gc.Filter(keys[startIdx:], vals[startIdx:]); !gcTS.Equal(proto.ZeroTimestamp) {
+						gcArgs.Keys = append(gcArgs.Keys, proto.InternalGCRequest_GCKey{Key: expBaseKey, Timestamp: gcTS})
```

Comment:
I'm concerned about the potential size of gcArgs.Keys. For ranges consisting mostly of stats the set of keys to be GC'd could end up being a large fraction of the range size, and that's expensive to pass through raft in one blob. We should consider splitting up GC requests.